### PR TITLE
Clarity character stats

### LINKED ISSUE #8994: Bring back ability cooldown info
https://github.com/Database-Clarity/Character-Stats

### PR DESCRIPTION
**Pull request by bhollis**
Fixes #8994. This was a longer road than I thought. I've managed to make some progress but wanted to stop here to get feedback on the direction.

* We load cooldowns data from the Clarity database. 
* Right now I have the individual ability cooldowns tied to the active subclass. Otherwise it ends up as big lists of like every possible grenade with their cooldowns.
* I'm not sure whether to just show the cooldowns, or to show current cooldown plus the value at the next tier (like the game does), or to try and display a fun little graph:

Screenshot 2023-05-11 at 7 08 27 PM
Screenshot 2023-05-11 at 7 08 32 PM
Screenshot 2023-05-11 at 7 08 35 PM
Screenshot 2023-05-11 at 7 08 39 PM
Screenshot 2023-05-11 at 7 08 42 PM
Screenshot 2023-05-11 at 7 08 46 PM

TODO:

* [x] The subsection in the tooltip should be colored/labeled "Community Insight" like the perk descriptions are.
* [x] Include exotic-perk and aspect overrides to cooldowns.
* [x] Styling, cleanup, disentangling the various loadouts paths.

@StardustD2, would like your opinion on what's most useful. @vivekhnz would love your ideas on presentation as well.

### CODE REVIEW

**Comment by StardustD2** [comment]
First off, I'm a fan of how this turned out but it might be a bit too much. Don't really want to comment on UI/UX as I'm notoriously terrible at that.

Apart from that I have a few things to note:
- I would generally recommend displaying rounded numbers when it comes to ability cooldowns as unfortunately the method I use to calculate them isn't *that* precise. The only reason I ended up rounding to 2 decimal places is because it improves accuracy when combined with cooldown overrides.
- Unsure if worth implementing for you but I did end up adding a bit of error handling with `update.json` in case I make a breaking change (which I will be doing again soon, more on that later). So that way it would have a fallback directory with a file it would be able to handle just fine.
- The graphs seem to only have 10 bars with info about T10 of the Character Stat missing?
- If a character stat does multiple things, I think it would make sense for its influence on ability cooldowns to have priority and be placed at the top.

As for the breaking change I'll be implementing shortly, I'll be removing the CooldownOverride member of Overrides since that functionality is handled by Scalars anyway. Additionally, I'll be adding a new property to abilities themselves since Bungie is now messing with ability charge-based regeneration scalars when it comes to strand melees. Will follow up on these changes when I get around to implementing them.

**Comment by bhollis** [comment]
Yeah the graphs are not a final idea :-)

**Comment by StardustD2** [comment]
Quick note since the commit notification email woke me up: I'd suggest only rounding the actual cooldowns themselves as the regeneration times, walk speeds, and other miscellaneous stats that aren't shown in-game are from precise testing. Rounding (in my opinion) is counterproductive with those.

**Comment by vivekhnz** [comment]
Nice, this seems like it'll be useful.

Some thoughts on presentation:
- I don't think the graphs are necessary.
  - If we want to convey the relative impact of a tier compared to the baseline, I think the % improvement compared to T1 would be a more compact way to display this info.
  - The graphs do a good job of illustrating non-linear stat tier benefits, but I'm not sure how useful this information is. Furthermore, it seems like Bungie has been moving towards linearising most of the stats.
- The values might be easier to scan if we:
  - Bold them
  - Horizontally align stats, like we do in the item popup
  - Format duration values to include minutes i.e. `126s -> 2m 4s`
- Displaying the values at every tier would be overkill, but I think displaying `T1`, `T(current-1)`, `T(current)`, `T(current+1)` and `T10` would be useful.
  - `T1` tells me the impact if I don't invest anything into this stat.
  - `T10` tells me the best I can get if I fully invest into this stat.
  - `T(current-1)` helps me understand how important it is to maintain my current stat tier.
  - `T(current+1)` helps me understand the potential benefit of investing a little bit more into this stat.
  - I think all of these are useful things to know, especially when building a loadout.

Here's a few mockups (not particularly attached to any of them):
dim-character-stat-mockups

- The tooltips in loadout optimizer are displaying too many ability values at the moment.
  - We probably need to filter abilities based on selected class, then group by name or similar (to handle duplicated class abilities across every subclass), then narrow down further based on the user's selected subclass.

image

As a side-note, it might be really useful to expose these values near the min & max stat tier selectors in Loadout Optimizer e.g.
dim-loadout-optimizer-stat-tiers

**Comment by bhollis** [comment]
Great feedback. I believe the loadout optimizer stuff just isn't hooked up to the selected subclass yet but I'll check.

**Comment by StardustD2** [comment]
Finally found some time to work on the repo and update the schema and generator code with the necessary changes.
Still working on some things but wanted to make a PR in the meantime in case you want to check out how things will work/look going forward. [Pr here.](https://github.com/Database-Clarity/Character-Stats/pull/3)

**Comment by bhollis** [comment]
OK, I think this is pretty much ready for review. I haven't done some more advanced stuff like falling back to the current subclass when loadouts don't specify a subclass, but otherwise I think things are pretty clear:

Screenshot 2023-05-25 at 3 26 41 PM

**Comment by StardustD2** [comment]
I'm curious about one specific edge case and how that would end up looking like. Namely, Void Towering Barricade with Bastion Void Aspect and Citan's Ramparts Exotic Arms together.
It's the one multi-stacking override example I can think of.

**Comment by bhollis** [comment]
It looks like this:

Screenshot 2023-05-25 at 3 51 06 PM

**Comment by StardustD2** [comment]
That looks about right to me. I love how everything turned out. It's everything I could've hoped for and more. Wonderful job!

**Comment by StardustD2** [comment]
Oh right one quick note: for resilience be sure to rename that to PVE damage resist or something of the sort before this goes to prod. Definitely don't want tl confuse people.

**Review by robojumper — COMMENTED** [review]
The overrides mostly wrap to their own line, but for short names they can appear on the same line -- maybe we can force them to ~~wrap~~ stack vertically?

grafik

File: src/app/store-stats/ClarityCharacterStat.m.scss [hunk 1]
```diff
@@ -0,0 +1,68 @@
+@use '../variables' as *;
+@use '../dim-ui/tooltip-mixins' as *;
+
+.communityInsightSection {
+  @include tooltip-section-color($communityBlue);
```
Comment (vivekhnz) [inline src/app/store-stats/ClarityCharacterStat.m.scss:5]: This mixin should be removing the tooltip section's top border, but isn't.

dim-tooltip-section-top-border

Perhaps we tack on a `!important` [here](https://github.com/DestinyItemManager/DIM/blob/clarity-character-stats/src/app/dim-ui/_tooltip-mixins.scss#L22) to fix that?

```diff
-border-top-color: transparent;
+border-top-color: transparent !important;
```

File: src/app/store-stats/ClarityCharacterStat.m.scss [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  .currentColumn {
+    opacity: 1;
+  }
+}
+
+.unit {
+  font-weight: normal !important;
+  font-size: 0.85em;
```
Comment (vivekhnz) [inline src/app/store-stats/ClarityCharacterStat.m.scss:62]:
```suggestion
  font-size: 0.85em;
  line-height: 1em;
```
This should keep the units on the same baseline as the value.

File: src/app/store-stats/ClarityCharacterStat.tsx [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        ))}
+      </th>
+      {tier - 1 >= 0 && (
+        <>
+          <td>{cooldowns[tier - 1].toLocaleString()}</td>
+          {unitEl}
+        </>
+      )}
+      <td className={styles.currentColumn}>{cooldowns[tier].toLocaleString()}</td>
+      {unitEl}
```
Comment (vivekhnz) [inline src/app/store-stats/ClarityCharacterStat.tsx:238]: The table might be easier to scan if the units also get the current column styling.

dim-stat-tooltip-unit-color

File: src/app/store-stats/StatTooltip.tsx [hunk 4]
```diff
@@ -13,8 +16,23 @@ interface Stat {
...
+  stat: Stat;
+  /**
+   * Hashes of equipped/selected items and subclass plugs for this character or loadout. Can be limited to
+   * exotic armor + subclass plugs - make sure to include default-selected subclass plugs.
+   */
+  equippedHashes: Set<number>;
+}) {
   const tier = statTier(stat.value);
+  const descriptionsToDisplay = useSelector(settingSelector('descriptionsToDisplay'));
+  const useClarityInfo = descriptionsToDisplay !== 'bungie';
```
Comment (vivekhnz) [inline src/app/store-stats/StatTooltip.tsx:35]: Should we also check the Clarity descriptions feature flag here?

```suggestion
  const useClarityInfo = $featureFlags.clarityDescriptions && descriptionsToDisplay !== 'bungie';
```

Comment (vivekhnz) [reply src/app/store-stats/StatTooltip.tsx:35]: Also, should we hide the Bungie-provided stat description if the user has set `Community Perk Insights` to `Community Descriptions`?

**Review by vivekhnz — APPROVED** [review]
Looks great! I've left some suggestions inline.

File: src/app/store-stats/StatTooltip.tsx (hunk 4 above)
Comment (StardustD2) [reply src/app/store-stats/StatTooltip.tsx:35]: Probably not for now.
I do plan to add a "community description" to the stats themselves at a later date and maybe you could do it at that point?

**Comment by bhollis** [comment]
Thanks for the feedback, I've made those changes.

**Comment by vivekhnz** [comment]
Looking really good!

One small thing I noticed - for abilities with long names (e.g. `Shadowshot: Moebius Quiver`), the table can sometimes extend beyond the tooltip's bounds.

image

**Pull request merged**